tests/auto/layouts/tst_layouts.cpp

```diff
@@ -0,0 +1,230 @@
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License along
+  with this program; if not, write to the Free Software Foundation, Inc.,
+  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
+#include <QtTest>
+
+#include "hboxlayoutwidget.h"
+#include "vboxlayoutwidget.h"
+#include "formlayoutwidget.h"
+#include "gridlayoutwidget.h"
+#include "stackedlayoutwidget.h"
+#include "stackedwidget.h"
+
+#include "declarativewidgetsdocument.h"
+
+#include <QLayout>
+
+typedef std::shared_ptr<QWidget> QWidgetPtr;
+typedef std::shared_ptr<DeclarativeWidgetsDocument> DeclarativeWidgetsDocumentPtr;
+
+class tst_Layouts : public QObject
+{
+    Q_OBJECT
+public:
+    tst_Layouts();
+
+private slots:
+    void hBoxLayout();
+    void vBoxLayout();
+    void formLayout();
+    void gridLayout();
+    void stackedLayout();
+    void stackedWidget();
+
+private:
+    void testLayouts(QWidgetPtr uiWidget, DeclarativeWidgetsDocumentPtr declarativeDocument);
+    void compareLayouts(QLayout *a, QLayout *b);
+    void compareLayoutItems(QLayoutItem* a, QLayoutItem* b);
+    void compareWidgets(QWidget* a, QWidget* b);
+    void compareGeometry(const QRect& a, const QRect& b);
+};
+
+tst_Layouts::tst_Layouts()
+    : QObject()
+{
+
+}
+
+void tst_Layouts::hBoxLayout()
+{
+    testLayouts( QWidgetPtr(new HBoxLayoutWidget())
+                 , DeclarativeWidgetsDocumentPtr(new DeclarativeWidgetsDocument(QStringLiteral("qrc:/qml/HBoxLayoutTest.qml"))));
+}
+
+void tst_Layouts::vBoxLayout()
+{
+    testLayouts( QWidgetPtr(new VBoxLayoutWidget())
+                 , DeclarativeWidgetsDocumentPtr(new DeclarativeWidgetsDocument(QStringLiteral("qrc:/qml/VBoxLayoutTest.qml"))));
+}
+
+void tst_Layouts::formLayout()
+{
+    testLayouts( QWidgetPtr(new FormLayoutWidget())
+                 , DeclarativeWidgetsDocumentPtr(new DeclarativeWidgetsDocument(QStringLiteral("qrc:/qml/FormLayoutTest.qml"))));
+}
+
+void tst_Layouts::gridLayout()
+{
+    testLayouts( QWidgetPtr(new GridLayoutWidget())
+                 , DeclarativeWidgetsDocumentPtr(new DeclarativeWidgetsDocument(QStringLiteral("qrc:/qml/GridLayoutTest.qml"))));
+}
+
+void tst_Layouts::stackedLayout()
+{
+    testLayouts( QWidgetPtr(new StackedLayoutWidget())
+                 , DeclarativeWidgetsDocumentPtr(new DeclarativeWidgetsDocument(QStringLiteral("qrc:/qml/StackedLayoutTest.qml"))));
+}
+
+void tst_Layouts::stackedWidget()
+{
+    testLayouts( QWidgetPtr(new StackedWidget())
+                 , DeclarativeWidgetsDocumentPtr(new DeclarativeWidgetsDocument(QStringLiteral("qrc:/qml/StackedWidgetTest.qml"))));
+}
+
+void tst_Layouts::testLayouts(QWidgetPtr uiWidget, DeclarativeWidgetsDocumentPtr declarativeDocument)
+{
+    QVERIFY(uiWidget != nullptr);
+    QVERIFY(declarativeDocument != nullptr);
+
+    QWidgetPtr declarativeWidget(declarativeDocument->create<QWidget>());
```

Comment:
I would say we switch this to use QQmlComponent for loading and creating the DeclarativeWidget tree.
Otherwise we will have to change this all over again when we deprecate DeclarativeWidgetsDocument in favor of the plugin